otherlibs/systhreads/st_stubs.c

```diff
@@ -703,9 +706,15 @@ CAMLexport int caml_c_thread_register(void)
     Tick_thread_running = 1;
   }
 
-  /* Release the master lock */
-  thread_lock_release(Dom_c_threads);
+  /* Release the domain lock the regular way. Note: we cannot receive
+     an exception here. */
+  caml_enter_blocking_section_no_pending();
```

Comment:
At this point there are many ways offered to "enter blocking section" or "release the runtime lock", including this one or `thread_lock_release` but also `caml_thread_enter_blocking_section`. I'm a bit lost as to which does what. Can you explain why `caml_enter_blocking_section(_no_pending)` is the right one to use here, and maybe include this explanation in the comment on lines 709-710?

Reply:
One thing that is confusing to me is that "thread lock" and "domain lock" are two notions that are not orthogonal but overlapping. In particular, if I understand correctly, `thread_lock_release` eventually calls `caml_release_domain_lock` (among other things).

Reading this code, I have no clear idea whether one of the following holds:
1. in the success path, the stuff that occurs before `caml_enter_blocking_section` sums up to the equivalent of what `thread_lock_release` would have done, so we end up with something that is equivalent to a well-bracketed acquisition and release of the thread lock, or
2. in the success path, the code is doing *less* than what `thread_lock_release` would have done, because the point is to keep ownership of *something* after the function returns

(My bet would be that (1) holds, that those `caml_c_{register,unregister}` functions are expected to be called from foreign threads that are not "active" in any respect with regard to the OCaml runtime (or equivalently, that can run concurrently with the runtime or OCaml mutators), and preserve this property on exit.)

Reply:
Indeed the two functions must be intimate of the details of the relationship the thread lock and the domain lock. The goal is to construct and destruct an "OCaml" thread from scratch which is why it deals with this the details of acquiring/releasing the lock. Out of the four locking operations, though, two should be straightforward. If the C thread is correctly attached to the domain, then one can call caml_enter_blocking_section at the end of having attached it, and caml_leave_blocking_section at the beginning of detaching it. Indeed the point is to have a thread with which caml_enter/leave_blocking_section are functional. I think the current comment meant to explain this is clear enough, but this might be more clear when reading the code directly rather than from the diff.